core/src/main/java/com/graphhopper/storage/BaseGraph.java

```diff
@@ -827,39 +828,35 @@ private PointList fetchWayGeometry_(long edgePointer, boolean reverse, FetchMode
             geoRef += 4L;
             bytes = new byte[count * nodeAccess.getDimension() * 4];
             wayGeometry.getBytes(geoRef, bytes, bytes.length);
-        } else if (mode == FetchMode.PILLAR_ONLY)
-            return PointList.EMPTY;
+        } else if (mode == FetchMode.PILLAR_ONLY) {
+            visitor.init(0, nodeAccess.is3D());
+            return;
+        }
 
-        PointList pillarNodes = new PointList(getPointListLength(count, mode), nodeAccess.is3D());
-        if (reverse) {
-            if (mode == FetchMode.ALL || mode == FetchMode.PILLAR_AND_ADJ)
-                pillarNodes.add(nodeAccess, adjNode);
-        } else if (mode == FetchMode.ALL || mode == FetchMode.BASE_AND_PILLAR)
-            pillarNodes.add(nodeAccess, baseNode);
+        visitor.init(getPointListLength(count, mode), nodeAccess.is3D());
+        if (mode == FetchMode.ALL || mode == FetchMode.BASE_AND_PILLAR)
+            visitPoint(nodeAccess, baseNode, visitor);
 
         int index = 0;
+        int pointOffset = (nodeAccess.is3D() ? 3 : 2) * 4;
+        if (reverse) {
+            index = (count-1) * pointOffset;
+            pointOffset = (-1) * pointOffset;
+        }
```

Comment:
The old code reversed the resulting List. Here we just iterate in reverse direction.